src/dodal/beamlines/new-i22.py

```diff
@@ -0,0 +1,305 @@
+from pathlib import Path
+
+from ophyd_async.panda import HDFPanda
+
+from dodal.common.beamlines.beamline_utils import (
+    get_directory_provider,
+    set_directory_provider,
+)
+from dodal.common.beamlines.beamline_utils import set_beamline as set_utils_beamline
+from dodal.common.beamlines.device_helpers import numbered_slits
+from dodal.common.beamlines.instantiation_behaviour import instantiation_behaviour
+from dodal.common.visit import DirectoryServiceClient, StaticVisitDirectoryProvider
+from dodal.devices.focusing_mirror import FocusingMirror
+from dodal.devices.i22.dcm import CrystalMetadata, DoubleCrystalMonochromator
+from dodal.devices.i22.fswitch import FSwitch
+from dodal.devices.i22.nxsas import NXSasMetadataHolder, NXSasOAV, NXSasPilatus
+from dodal.devices.linkam3 import Linkam3
+from dodal.devices.synchrotron import Synchrotron
+from dodal.devices.tetramm import TetrammDetector
+from dodal.devices.undulator import Undulator
+from dodal.log import set_beamline as set_log_beamline
+from dodal.utils import BeamlinePrefix, get_beamline_name, skip_device
+
+BL = get_beamline_name("i22")
+set_log_beamline(BL)
+set_utils_beamline(BL)
+
+# Currently we must hard-code the visit, determining the visit at runtime requires
+# infrastructure that is still WIP.
+# Communication with GDA is also WIP so for now we determine an arbitrary scan number
+# locally and write the commissioning directory. The scan number is not guaranteed to
+# be unique and the data is at risk - this configuration is for testing only.
+set_directory_provider(
+    StaticVisitDirectoryProvider(
+        BL,
+        Path("/dls/i22/data/2024/cm37271-2/bluesky"),
+        client=DirectoryServiceClient("http://i22-control:8088/api"),
+    )
+)
+
+
+@instantiation_behaviour(
+    default_use_mock_at_connection=True, default_timeout_for_connect=10
+)
+def saxs():
+    """Create a SAXS detector with specific settings."""
+    metadata_holder = NXSasMetadataHolder(
+        x_pixel_size=(1.72e-1, "mm"),
+        y_pixel_size=(1.72e-1, "mm"),
+        description="Dectris Pilatus3 2M",
+        type="Photon Counting Hybrid Pixel",
+        sensor_material="silicon",
+        sensor_thickness=(0.45, "mm"),
+        distance=(4711.833684146172, "mm"),
+    )
+
+    return NXSasPilatus(
+        name="saxs",
+        prefix="-EA-PILAT-01:",
+        drv_suffix="CAM:",
+        hdf_suffix="HDF5:",
+        metadata_holder=metadata_holder,
+        directory_provider=get_directory_provider(),
+    )
+
+
+@instantiation_behaviour(
+    default_use_mock_at_connection=True, default_timeout_for_connect=10
+)
+def synchrotron():
```

Comment:
Should: So the idea is that this can still be called with `i22.sychrotron(mock=False, timeout=100)` if you want but if you don't specify them you get the defaults? If so it's really not clear that this is possible as:
1. There are no tests covering the behavior
2. A naïve glance at this function (i.e. from a scientist's perspective) makes it look like it takes no arguments at all. Could we instead use the function definition `def synchrotron(mock=None, timeout=None)`? This still looks weird as it looks like the parameters aren't being used but at least it hints there's something wacky going on.

Reply:
do you mean in a context different from `blueapi`?

in the plan repos I've seen the 'inject' operator used most often.

I did not consider IPython terminal as a use case.

Reply:
> Yes, calling these factories directly, for example from a terminal, is a usecase that we should consider.

do we have a list of use cases to be considered? afaik this was not in the initial spec for the issue and neither can I find a reference to it in the repository at the moment. I was under the impression for the core team the the use of context of the plans and devices is through plans accessible through blueapi.

Reply:
That's how the DAQ core team use it, but Controls core (i.e. Tom) and MX are likely to want to instantiate the device directly for debugging.

@device_factory()

should mutate a Callable[[P], Device] with P = ParamSpec("P") to Callable[[Concatenate[P, bool, bool, float, bool, bool]], Device] where each of the additional calls is a kw only arg of:

```
    eager: bool = True,
    set_name: bool = True,
    default_timeout: float = DEFAULT_TIMEOUT,
    default_mock: bool = False,
    default_skip: bool = False
```

so for something like

```
@device_factory()
def synchrotron():
    return Synchrotron()
```

actually becomes in the terminal

```
def synchrotron(eager: bool = True, set_name: bool = True, timeout: float = 10.0, mock: bool = False, skip: bool = False):
...
```

with those params being used by the DeviceInitializationController

Reply:
I've made a suggestion of what I think the decorator should do, now it turns

Callable[P, T] -> Callable[Concatenate[bool | None, bool | None, float | None, bool | None, _skip | None, P], T]

i.e. 

@device_factory()
def synchrotron() -> Synchrotron:
    return Synchrotron()

becomes 

@device_factory()
def synchrotron(eager: bool | None = None,
                name: str | None = None,
                timeout: float | None = None,
                mock: bool | None = None,
                skip: bool | Callable[[], bool] | None = None) -> Synchrotron | None:
""" If None, default arguments from the device_factory are used"""
    ...

Reply:
No, currently

```
@device_factory()
def synchrotron() -> Synchrotron:
    return Synchrotron()
```

becomes 

```
def synchrotron():
    ... does some handling with the default values
    return Synchrotron()
```

which doesn't allow overwriting the default values from the command line.

Reply:
Important to remember that 

```
@decorator
def function():
   return return_value
```

means that the decorator is run when the file is first imported. After `function` has been defined, it's automatically and immediately wrapped by `@decorator` and anyone trying to call `function(args)` is actually seeing/calling `decorator(function)(args)`

Which is what I mean by `becomes` in the above: we want the definition in our file to be simple, but the returned value to be complete and useful.

Reply:
I revisited this a little:

The attached turns

```python
@device_factory
def synchrotron():
    return Synchrotron()
```

into

```python
def synchrotron(eager: bool | None = None,
                name: str | None = None,
                timeout: float | None = None,
                mock: bool | None = None) -> Synchrotron:
```
Since presumably if you're calling the function, you would never want it to return None.
skipped is then a property of the `DeviceInitializationController` and blueapi (or similar) can get all the definitions from a file, filter to just the `DeviceInitializationController`, check controller.skipped, check if controller.device exists already and get the value as required.


[foo.txt](https://github.com/user-attachments/files/16785102/foo.txt)